File: 백준/Gold/2458. 키 순서/키 순서.py
import sys
input = sys.stdin.readline

n, m = map(int, input().split())
graph = [[0 for _ in range(n)]for _ in range(n)]
for _ in range(m):
    s, e = map(int, input().split())
    graph[s-1][e-1]=1
#앞, 뒤 모두 있는지 확인(연결여부)
for path in range(n):
    for i in range(n):
        for j in range(n):
            if graph[i][path] == 1 and graph[path][j] == 1:
                graph[i][j] =1
res = 0
for i in range(n):
    check = 0
    for j in range(n):
        check += (graph[i][j] + graph[j][i]) #내 앞, 뒤가 모두 있을 경우
    # 자기 제외 모두 체크
    if check == n - 1:
        res +=1
print(res)

File: 백준/Gold/2458. 키 순서/키 순서.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

var (
	reader = bufio.NewReader(os.Stdin)
	writer = bufio.NewWriter(os.Stdout)
)

func main() {
	defer writer.Flush()
	var n, m, s, e, res, check int
	fmt.Fscanf(reader, "%d %d\n", &n, &m)
	var graph = make([][]int, n)
	for i := 0; i < n; i++ {
		graph[i] = make([]int, n)
	}
	for i := 0; i < m; i++ {
		fmt.Fscanf(reader, "%d %d\n", &s, &e)
		graph[s-1][e-1] = 1
	}
	for path := 0; path < n; path++ {
		for i := 0; i < n; i++ {
			for j := 0; j < n; j++ {
				if graph[i][path] == 1 && graph[path][j] == 1 {
					graph[i][j] = 1
				}
			}
		}
	}
	for i := 0; i < n; i++ {
		check = 0
		for j := 0; j < n; j++ {
			check += (graph[i][j] + graph[j][i])
		}
		if check == n-1 {
			res++
		}
	}
	fmt.Fprintf(writer, "%d\n", res)
}
